Remove runtime functions from the stacktrace when runtime built with -g

With #964 if the runtime is compiled with `-g` stack trace in the case of a panic will include the runtime functions (since runtime function calls now have locations). For example in the case of `01-int/div1-p.bal` we will get the stack trace as
```
panic: arithmetic overflow
    _bal_error_construct /home/heshan/work/nballerina/runtime/error.c:10
    _bal_panic_construct /home/heshan/work/nballerina/runtime/panic.c:67
    div ../../../compiler/testSuite/01-int/div1-p.bal
    main (root) ../../../compiler/testSuite/01-int/div1-p.bal:6
    main /home/heshan/work/nballerina/runtime/main.c:11
```
instead of 
```
panic: arithmetic overflow
    div ../../../compiler/testSuite/01-int/div1-p.bal:10
    main (root) ../../../compiler/testSuite/01-int/div1-p.bal:6
```

--- a/runtime/error.c
+++ b/runtime/error.c
@@ -61,6 +61,7 @@ static void processInitialPC(PC pc, BacktraceStartLine *backtraceStartLine);
 static enum DemangleResult demangle(const char *mangledName, const char **localName, FILE *fp);
 static bool demangleModules(const char **mangledModules, FILE *fp);
 static bool demangleCountedName(const char **pp, const char **name);
+static bool isRuntimeFunction(const char* mangledFunctionName);
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
     SimpleBacktrace simpleBacktrace;
@@ -176,7 +177,6 @@ static void processPCs(GC PC* pcs, uint32_t nPCs, uint32_t start, int64_t lineNu
     // Handle the starting pc seperately
     BacktraceStartLine backtraceStartLine = { { error->code, error->message, error->fp }, lineNumber };
     processInitialPC(pcs[start], &backtraceStartLine);
-    backtraceStartLine.lineNumber = 0;
     if (backtraceStartLine.error.code == NO_DEBUG_INFO) {
         for (uint32_t i = start + 1; i < nPCs; i++) {
             // We do not need to pass error_callback because, already we have an error.
@@ -203,8 +203,13 @@ static void processInitialPC(PC pc, BacktraceStartLine *backtraceStartLine) {
 // Implementation of backtrace_full_callback
 static int printBacktraceLineCB(void *data, UNUSED PC pc, const char *filename, int lineno, const char *function) {
     BacktraceStartLine *backtraceStartLine = data;
-    int64_t lineNumber = backtraceStartLine->lineNumber != 0 ? backtraceStartLine->lineNumber : lineno;
-    printBacktraceLine(filename, lineNumber, function, backtraceStartLine->error.fp);
+    if (function != NULL && !isRuntimeFunction(function)) {
+        int64_t lineNumber = backtraceStartLine->lineNumber != 0 ? backtraceStartLine->lineNumber : lineno;
+        if (backtraceStartLine->lineNumber != 0) {
+            backtraceStartLine->lineNumber = 0;
+        }
+        printBacktraceLine(filename, lineNumber, function, backtraceStartLine->error.fp);
+    }
     return SUCCESS;
 }
 
@@ -233,6 +238,15 @@ static void printBacktraceLine(const char *filename, int64_t lineNumber, const c
         putc('\n', fp);
     }
     fflush(fp);
+}
+
+static bool isRuntimeFunction(const char* mangledFunctionName) {
+    const char *runtimePre = "_bal_\0";
+    const char *libPre = "_Bb\0";
+    const char *initMain = "main\0";
+    return strncmp(runtimePre, mangledFunctionName, strlen(runtimePre)) == 0 ||
+           strncmp(libPre, mangledFunctionName, strlen(libPre)) == 0 ||
+           strcmp(initMain, mangledFunctionName) == 0;
 }
 
 // Implementation of backtrace_error_callback

--- a/runtime/error.c
+++ b/runtime/error.c
@@ -61,10 +61,7 @@ static void processInitialPC(PC pc, BacktraceStartLine *backtraceStartLine);
 static enum DemangleResult demangle(const char *mangledName, const char **localName, FILE *fp);
 static bool demangleModules(const char **mangledModules, FILE *fp);
 static bool demangleCountedName(const char **pp, const char **name);
-static bool isRuntimeFunction(const char* mangledFunctionName);
 static bool isPrefix(const char* prefix, const char* str);
-static bool isStringEq(const char* str1, const char* str2);
-static bool stringEq(const char* str1, const char* str2, int len);
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
     SimpleBacktrace simpleBacktrace;
@@ -206,7 +203,7 @@ static void processInitialPC(PC pc, BacktraceStartLine *backtraceStartLine) {
 // Implementation of backtrace_full_callback
 static int printBacktraceLineCB(void *data, UNUSED PC pc, const char *filename, int lineno, const char *function) {
     BacktraceStartLine *backtraceStartLine = data;
-    if (function != NULL && !isRuntimeFunction(function)) {
+    if (function != NULL && isPrefix("_B\0", function)) {
         int64_t lineNumber = backtraceStartLine->lineNumber != 0 ? backtraceStartLine->lineNumber : lineno;
         if (backtraceStartLine->lineNumber != 0) {
             backtraceStartLine->lineNumber = 0;
@@ -243,32 +240,13 @@ static void printBacktraceLine(const char *filename, int64_t lineNumber, const c
     fflush(fp);
 }
 
-static bool isRuntimeFunction(const char* mangledFunctionName) {
-    const char *runtimePre = "_bal_\0";
-    const char *libPre = "_Bb\0";
-    const char *initMain = "main\0";
-    return isPrefix(runtimePre, mangledFunctionName) || isPrefix(libPre, mangledFunctionName) || isStringEq(initMain, mangledFunctionName);
-}
-
 static bool isPrefix(const char* prefix, const char* str) {
     unsigned long len = strlen(prefix);
     if (strlen(str) < len) {
         return false;
     }
-    return stringEq(prefix, str, len);
-}
-
-static bool isStringEq(const char* str1, const char* str2) {
-    unsigned long len = strlen(str1);
-    if (strlen(str2) != len) {
-        return false;
-    }
-    return stringEq(str1, str2, len);
-}
-
-static bool stringEq(const char* str1, const char* str2, int len) {
     while (--len > 0) {
-        if (*++str1 != *++str2)
+        if (*++prefix != *++str)
             return false;
     }
     return true;

--- a/runtime/error.c
+++ b/runtime/error.c
@@ -241,11 +241,7 @@ static void printBacktraceLine(const char *filename, int64_t lineNumber, const c
 }
 
 static bool isPrefix(const char *prefix, const char *str) {
-    while (*++prefix == *++str) {
-        if (*prefix == '\0') {
-            break;
-        }
-    }
+    for (; *prefix == *str && *prefix != '\0'; prefix++, str++) { }
     return *prefix == '\0';
 }
 

--- a/runtime/error.c
+++ b/runtime/error.c
@@ -61,7 +61,7 @@ static void processInitialPC(PC pc, BacktraceStartLine *backtraceStartLine);
 static enum DemangleResult demangle(const char *mangledName, const char **localName, FILE *fp);
 static bool demangleModules(const char **mangledModules, FILE *fp);
 static bool demangleCountedName(const char **pp, const char **name);
-static bool isPrefix(const char* prefix, const char* str);
+static bool isPrefix(const char *prefix, const char *str);
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
     SimpleBacktrace simpleBacktrace;
@@ -240,16 +240,13 @@ static void printBacktraceLine(const char *filename, int64_t lineNumber, const c
     fflush(fp);
 }
 
-static bool isPrefix(const char* prefix, const char* str) {
-    unsigned long len = strlen(prefix);
-    if (strlen(str) < len) {
-        return false;
+static bool isPrefix(const char *prefix, const char *str) {
+    while (*++prefix == *++str) {
+        if (*prefix == '\0') {
+            break;
+        }
     }
-    while (--len > 0) {
-        if (*++prefix != *++str)
-            return false;
-    }
-    return true;
+    return *prefix == '\0';
 }
 
 // Implementation of backtrace_error_callback

--- a/runtime/error.c
+++ b/runtime/error.c
@@ -62,6 +62,9 @@ static enum DemangleResult demangle(const char *mangledName, const char **localN
 static bool demangleModules(const char **mangledModules, FILE *fp);
 static bool demangleCountedName(const char **pp, const char **name);
 static bool isRuntimeFunction(const char* mangledFunctionName);
+static bool isPrefix(const char* prefix, const char* str);
+static bool isStringEq(const char* str1, const char* str2);
+static bool stringEq(const char* str1, const char* str2, int len);
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
     SimpleBacktrace simpleBacktrace;
@@ -244,9 +247,31 @@ static bool isRuntimeFunction(const char* mangledFunctionName) {
     const char *runtimePre = "_bal_\0";
     const char *libPre = "_Bb\0";
     const char *initMain = "main\0";
-    return strncmp(runtimePre, mangledFunctionName, strlen(runtimePre)) == 0 ||
-           strncmp(libPre, mangledFunctionName, strlen(libPre)) == 0 ||
-           strcmp(initMain, mangledFunctionName) == 0;
+    return isPrefix(runtimePre, mangledFunctionName) || isPrefix(libPre, mangledFunctionName) || isStringEq(initMain, mangledFunctionName);
+}
+
+static bool isPrefix(const char* prefix, const char* str) {
+    unsigned long len = strlen(prefix);
+    if (strlen(str) < len) {
+        return false;
+    }
+    return stringEq(prefix, str, len);
+}
+
+static bool isStringEq(const char* str1, const char* str2) {
+    unsigned long len = strlen(str1);
+    if (strlen(str2) != len) {
+        return false;
+    }
+    return stringEq(str1, str2, len);
+}
+
+static bool stringEq(const char* str1, const char* str2, int len) {
+    while (--len > 0) {
+        if (*++str1 != *++str2)
+            return false;
+    }
+    return true;
 }
 
 // Implementation of backtrace_error_callback

--- a/test/runcheck.sh
+++ b/test/runcheck.sh
@@ -16,7 +16,7 @@ mkdir -p "$parent/actual"
 if test $kind == p; then
     balFile=`echo "$b" | sed -e 's/.exe/.bal/'`
     $("./$1" >/dev/null 2>"$out")
-    head -n 2 $out | sed -e 's/.*bal:/'"$balFile"':/' | diff -u "$2" -
+    grep -v ballerina/lang $out | head -n 2 | sed -e 's/.*bal:/'"$balFile"':/' | diff -u "$2" -
 else
     "./$1" >"$out"
     diff -u "$2" "$out"

